core/src/reservation.rs

```diff
@@ -0,0 +1,188 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Gas reservation structures.
+
+use crate::ids::{MessageId, ReservationId};
+use alloc::collections::BTreeMap;
+use codec::{Decode, Encode};
+use gear_core_errors::ExecutionError;
+use scale_info::TypeInfo;
+
+/// Gas reserver.
+///
+/// Controls gas reservations states.
+#[derive(Debug, Clone)]
+pub struct GasReserver {
+    message_id: MessageId,
+    nonce: u64,
+    states: GasReservationStates,
+    max_reservations: u64,
+}
+
+impl GasReserver {
+    /// Creates a new gas reserver.
+    pub fn new(
+        message_id: MessageId,
+        nonce: u64,
+        map: GasReservationMap,
+        max_reservations: u64,
+    ) -> Self {
+        Self {
+            message_id,
+            nonce,
+            states: map
+                .into_iter()
+                .map(|(id, GasReservationSlot { amount, expiration })| {
+                    (id, GasReservationState::Exists { amount, expiration })
+                })
+                .collect(),
+            max_reservations,
+        }
+    }
+
+    /// Reserves gas.
+    pub fn reserve(&mut self, amount: u64, duration: u32) -> Result<ReservationId, ExecutionError> {
+        let current_reservations = self
+            .states
+            .values()
+            .map(|state| {
+                matches!(
+                    state,
+                    GasReservationState::Exists { .. } | GasReservationState::Created { .. }
```

Comment:
seems that we should `exists.len()` + `created.len()` - `removed.len()` here

Reply:
or store somewhere this 3 numerics to optimise iterations - that's really expensive for future